File: haskell/src/Solutions/Euler139.hs
module Solutions.Euler139 where

-- https://en.wikipedia.org/wiki/Tree_of_primitive_Pythagorean_triples
matrixA (a, b, c) = (a - 2 * b + 2 * c, 2 * a - b + 2 * c, 2 * a - 2 * b + 3 * c)
matrixB (a, b, c) = (a + 2 * b + 2 * c, 2 * a + b + 2 * c, 2 * a + 2 * b + 3 * c)
matrixC (a, b, c) = (-a + 2 * b + 2 * c, -2 * a + b + 2 * c, -2 * a + 2 * b + 3 * c)

bound = 10^8 - 1
count t@(a, b, c)
    | per > bound = 0
    | otherwise = s + sum (map (count . ($t)) [matrixA, matrixB, matrixC])
    where per = a + b + c
          s = if c `mod` (b - a) == 0 then bound `div` per else 0

euler139 :: IO String
euler139 = return $ show $ count (3, 4, 5)

File: rust/src/solutions/euler139.rs
type P = (i64, i64, i64);

// https://en.wikipedia.org/wiki/Tree_of_primitive_Pythagorean_triples
fn matrix_a((a, b, c): P) -> P {
    (a - 2 * b + 2 * c, 2 * a - b + 2 * c, 2 * a - 2 * b + 3 * c)
}

fn matrix_b((a, b, c): P) -> P {
    (a + 2 * b + 2 * c, 2 * a + b + 2 * c, 2 * a + 2 * b + 3 * c)
}

fn matrix_c((a, b, c): P) -> P {
    (
        -a + 2 * b + 2 * c,
        -2 * a + b + 2 * c,
        -2 * a + 2 * b + 3 * c,
    )
}

const BOUND: i64 = 100000000 - 1;

fn count(t: P) -> i64 {
    let (a, b, c) = t;
    let per = a + b + c;
    if per > BOUND {
        0
    } else {
        let s = if c % (b - a) == 0 { BOUND / per } else { 0 };
        vec![count(matrix_a(t)), count(matrix_b(t)), count(matrix_c(t))]
            .into_iter()
            .sum::<i64>()
            + s
    }
}

pub fn euler139() -> String {
    count((3, 4, 5)).to_string()
}
